execution/types/runtime_decoder_test.go

```diff
@@ -0,0 +1,343 @@
+	decodeBlockBytes := append(lenExtrinsics, buffExtrinsic.Bytes()...)
+	decodeBlockBytes = append(header.Bytes(), decodeBlockBytes...)
+
+	mockCallOne.On("Encode", mock.Anything)
+
+	decodeBlockBuff := bytes.NewBuffer(decodeBlockBytes)
+
+	mockModuleOne.On("GetIndex").Return(moduleOneIdx)
+	mockSignedExtra.On("Decode", mock.Anything).Return()
+
+	assert.PanicsWithValue(t, "module with index ["+strconv.Itoa(int(idxModuleNotExists))+"] not found", func() {
+		target.DecodeBlock(decodeBlockBuff)
+	})
+}
+
+func Test_RuntimeDecoder_DecodeBlock_Function_Not_Exists(t *testing.T) {
+	target := setupRuntimeDecoder()
+
+	moduleFunctions[0] = mockCallOne
+
+	idxFunctionNotExists := sc.U8(10)
+
+	nonExistentExtrinsicsBytes := []byte{
+		132,                                                                                               // version
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // signer
+		0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+		1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+		1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+		1, 1, 1, 1, 1, 1, 1, // signature,
+		// extra
+		uint8(moduleOneIdx), uint8(idxFunctionNotExists), // call
+	}
+
+	buffExtrinsic := bytes.NewBuffer(append(sc.ToCompact(len(nonExistentExtrinsicsBytes)).Bytes(), nonExistentExtrinsicsBytes...))
+
+	lenExtrinsics := sc.ToCompact(1).Bytes()
+	decodeBlockBytes := append(lenExtrinsics, buffExtrinsic.Bytes()...)
+	decodeBlockBytes = append(header.Bytes(), decodeBlockBytes...)
+
+	mockCallOne.On("Encode", mock.Anything)
+
+	decodeBlockBuff := bytes.NewBuffer(decodeBlockBytes)
+
+	mockModuleOne.On("GetIndex").Return(moduleOneIdx)
+	mockSignedExtra.On("Decode", mock.Anything).Return()
+	mockModuleOne.On("Functions").Return(moduleFunctions)
+
+	assert.PanicsWithValue(t, "function index ["+strconv.Itoa(int(idxFunctionNotExists))+"] for module ["+strconv.Itoa(int(moduleOneIdx))+"] not found", func() {
+		target.DecodeBlock(decodeBlockBuff)
+	})
+}
+
+func Test_RuntimeDecoder_DecodeBlock_Single_Extrinsic(t *testing.T) {
+	target := setupRuntimeDecoder()
+
+	moduleFunctions[0] = mockCallOne
+
+	buffExtrinsic := bytes.NewBuffer(append(sc.ToCompact(len(signedExtrinsicBytes)).Bytes(), signedExtrinsicBytes...))
+
+	lenExtrinsics := sc.ToCompact(1).Bytes()
+	decodeBlockBytes := append(lenExtrinsics, buffExtrinsic.Bytes()...)
+	decodeBlockBytes = append(header.Bytes(), decodeBlockBytes...)
+
+	mockCallOne.On("Encode", mock.Anything)
+
+	decodeBlockBuff := bytes.NewBuffer(decodeBlockBytes)
+
+	mockModuleOne.On("GetIndex").Return(moduleOneIdx)
+	mockSignedExtra.On("Decode", mock.Anything).Return()
+	mockModuleOne.On("Functions").Return(moduleFunctions)
+	mockCallOne.On("DecodeArgs", decodeBlockBuff).Return(mockCallOne)
+	result := target.DecodeBlock(decodeBlockBuff)
+
+	mockCallOne.On("DecodeArgs", buffExtrinsic).Return(mockCallOne)
+
+	decodedExtr := target.DecodeUncheckedExtrinsic(buffExtrinsic)
```

Comment:
I don't think you should be calling this method manually, as it is not a mock and is being called as part of the DecodeBlock, which is what we are testing here.

One approach is to initialize a block with a single extrinsic, encode it into a buffer, and then attempt to call the DecodeBlock method with that buffer to test if the returned result matches the original block.

Also, you may want to add a separate test for the DecodeUncheckedExtrinsic method which is public and used in other places

```go
func Test_RuntimeDecoder_DecodeUncheckedExtrinsic(t *testing.T) {
	target := setupRuntimeDecoder()

	moduleFunctions[0] = mockCallOne

	buf := bytes.NewBuffer(
		append(sc.ToCompact(len(signedExtrinsicBytes)).Bytes(), signedExtrinsicBytes...),
	)

	mockSignedExtra.On("Decode", mock.Anything).Return()
	mockModuleOne.On("GetIndex").Return(moduleOneIdx)
	mockModuleOne.On("Functions").Return(moduleFunctions)
	mockCallOne.On("DecodeArgs", buf).Return(mockCallOne)

	decodedExtr := target.DecodeUncheckedExtrinsic(buf)

	mockSignedExtra.AssertCalled(t, "Decode", mock.Anything)
	mockModuleOne.AssertCalled(t, "GetIndex")
	mockModuleOne.AssertCalled(t, "Functions")
	mockCallOne.AssertCalled(t, "DecodeArgs", buf)
	assert.Equal(t, decodedExtr.IsSigned(), true)
}
```

Reply:
```go
	target := setupRuntimeDecoder()

	moduleFunctions[0] = mockCallOne

	buffExtrinsic := bytes.NewBuffer(append(sc.ToCompact(len(signedExtrinsicBytes)).Bytes(), signedExtrinsicBytes...))

	lenExtrinsics := sc.ToCompact(1).Bytes()
	decodeBlockBytes := append(lenExtrinsics, buffExtrinsic.Bytes()...)
	decodeBlockBytes = append(header.Bytes(), decodeBlockBytes...)

	decodeBlockBuff := bytes.NewBuffer(decodeBlockBytes)

	mockSignedExtra.On("Decode", mock.Anything).Return()
	mockModuleOne.On("GetIndex").Return(moduleOneIdx)
	mockModuleOne.On("Functions").Return(moduleFunctions)
	mockCallOne.On("DecodeArgs", decodeBlockBuff).Return(mockCallOne)

	result := target.DecodeBlock(decodeBlockBuff)
```
you can check the `unchecked_extrinsic_test.go` to see how to initialize a test extrinsic
```go
	extrinsics := sc.Sequence[primitives.UncheckedExtrinsic]{
		NewUncheckedExtrinsic( ... ),
	}
```
```go
	expectedBlock := NewBlock(header, extrinsics)

	assert.Equal(t, expectedBlock.Header(), result.Header())
	assert.Equal(t, expectedBlock.Extrinsics(), result.Extrinsics())
```